### DISCUSSION TITLE
How to set method after server importing xml?

### DISCUSSION THREAD

**Discussion by LuckyHouse** [FreeOpcUa/opcua-asyncio#567] [Q&A]
![image](https://user-images.githubusercontent.com/19794322/117943533-d2264600-b33e-11eb-8562-be0d0ac1385a.png)

Hello, i'm new to OPCUA. 
My xml contains method nodes. It seems normal after importing 
But when i call the method UpdateCameras， the error message shown above is displayed.
Is there a way to reset the method of one node?

**Comment by AndreasHeine** [comment]
"assuming you are the ua server"
is the method bound to a callback (python-function)?

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/af5a31c0db2b4f2469cdc7c169a0aa812b606152/asyncua/server/address_space.py#L514

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/af5a31c0db2b4f2469cdc7c169a0aa812b606152/asyncua/server/server.py#L633
```python
    def link_method(self, node, callback):
        """
        Link a python function to a UA method in the address space; required when a UA method has been imported
        to the address space via XML; the python executable must be linked manually
        :param node: UA method node
        :param callback: python function that the UA method will call
        """
        self.iserver.isession.add_method_callback(node.nodeid, callback)
```

**Reply by LuckyHouse** [reply]
Thanks for reply. 
At the first startup, I will bind the callback function to the node through the "add_method" method, and then export the xml.
For second startup, import xml, and then it seems to reset the callback function here? like "set_method"?

Here is part of my code 

```
        if os.path.exists(ConfigurationConst.NodeSettingPath):
            nodes = await self.server.import_xml(ConfigurationConst.NodeSettingPath)
            self.cameras_info = await self.server.nodes.objects.get_child([f"0:{DeviceInfoConst.Cameras}"])
            nodes_id = await self.server.nodes.objects.get_child([f"0:{FunctionConst.UpdateCameraList}"])
        else:
            self.cameras_info = await self.create_node_id(DeviceInfoConst.Cameras,
                                                          ConfigurationConst.Variable,
                                                          ua.Variant([[]],
                                                                     varianttype=ua.VariantType.String,
                                                                     is_array=True))

            await self.create_node_id(FunctionConst.UpdateCameraList,
                                      ConfigurationConst.Method,
                                      (self.update_camera_list,
                                       None, [ua.VariantType.Int16]))

            await self.create_node_id(FunctionConst.SaveSetting,
                                      ConfigurationConst.Method,
                                      (self.save_project_setting,
                                       None, [ua.VariantType.Int16]))

            await self.create_node_id(FunctionConst.SelectCameras,
                                      ConfigurationConst.Method,
                                      (self.select_cameras,
                                       [ua.VariantType.Int16], [ua.VariantType.Int16]))

            # self.server.nodes.objects.get_child()

            await self.init_event()
            await self.server.export_xml_by_ns(ConfigurationConst.NodeSettingPath,
                                               self.opc_setting[ConfigurationConst.URI])
```

**Reply by AndreasHeine** [reply]
bind (link_method) after import!

**Reply by swamper123** [reply]
Your intention is right. `add_method` binds the OPC UA Method with the python function.
But when you export the Node, you just export the information "there should be a method, with these parameters". The python function will not be exported. Also the "link" to that python function is missing. 
After importing the xml file, you have to link each OPC UA method to the python function again, so the server knows, what should be done again. 
A wise man told me once something like "there is a difference between logic and informationmodel".

**Comment by swamper123** [comment]
The xmlimporter just imports the "Node" itself, but not the logic behind it. The method has to be linked with this server method:

https://github.com/FreeOpcUa/opcua-asyncio/blob/84b4a60f39c9d71e4c22b9c036611783d03de13d/asyncua/server/server.py#L634

I don't know atm if the other method stuff is imported as well (like input parameters with names etc....you may have to test this).

**Reply by LuckyHouse** [reply]
Thanks. I will try.

**Reply by LuckyHouse** [reply]
Cool. It works. Thank you!
